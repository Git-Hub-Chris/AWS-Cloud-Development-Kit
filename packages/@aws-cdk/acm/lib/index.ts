export * from './certificate';
export * from './certificate-ref';

// The L1 library for AWS::CertificateManager:
export * from '../cfn/certificatemanager';
