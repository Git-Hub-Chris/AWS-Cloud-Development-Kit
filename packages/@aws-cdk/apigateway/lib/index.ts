// The L1 Library for AWS::ApiGateway:
export * from '../cfn/apigateway';
