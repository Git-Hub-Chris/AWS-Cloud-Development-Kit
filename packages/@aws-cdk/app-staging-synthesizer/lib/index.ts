export * from './default-staging-stack';
export * from './synthesizer';
