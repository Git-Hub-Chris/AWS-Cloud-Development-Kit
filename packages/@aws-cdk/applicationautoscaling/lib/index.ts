// The L1 Library for AWS::ApplicationAutoScaling:
export * from '../cfn/applicationautoscaling';
