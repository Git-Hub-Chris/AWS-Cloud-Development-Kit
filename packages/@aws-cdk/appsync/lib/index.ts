// The L1 Library for AWS::AppSync:
export * from '../cfn/appsync';
