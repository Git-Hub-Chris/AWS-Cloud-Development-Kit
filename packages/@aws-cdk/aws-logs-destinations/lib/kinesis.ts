import iam = require('@aws-cdk/aws-iam');
import kinesis = require('@aws-cdk/aws-kinesis');
import logs = require('@aws-cdk/aws-logs');
import { Construct } from '@aws-cdk/cdk';

/**
 * Use a Kinesis stream as the destination for a log subscription
 */
export class KinesisDestination implements logs.ILogSubscriptionDestination {
  constructor(private readonly stream: kinesis.IStream) {
  }

  public bind(scope: Construct, _sourceLogGroup: logs.ILogGroup): logs.LogSubscriptionDestination {
    // Following example from https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/SubscriptionFilters.html#DestinationKinesisExample
    // Create a role to be assumed by CWL that can write to this stream and pass itself.
    const role = new iam.Role(scope, 'CloudWatchLogsCanPutRecords', {
      assumedBy: new iam.ServicePrincipal(`logs.amazonaws.com`)
    });
    this.stream.grantWrite(role);
    role.grantPassRole(role);
    return { arn: this.stream.streamArn, role };
  }
}