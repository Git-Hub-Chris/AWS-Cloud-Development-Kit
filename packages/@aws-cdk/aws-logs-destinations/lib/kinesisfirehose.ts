import * as iam from '@aws-cdk/aws-iam';
import * as firehose from '@aws-cdk/aws-kinesisfirehose';
import * as logs from '@aws-cdk/aws-logs';

// keep this import separate from other imports to reduce chance for merge conflicts with v2-main
// eslint-disable-next-line no-duplicate-imports, import/order
import { Construct } from '@aws-cdk/core';

/**
 * Use a Kinesis Firehose as the destination for a log subscription
 */
export class KinesisFirehoseDestination implements logs.ILogSubscriptionDestination {
  constructor(private readonly deliveryStream: firehose.IDeliveryStream) {
  }

  public bind(scope: Construct, _sourceLogGroup: logs.ILogGroup): logs.LogSubscriptionDestinationConfig {
    const region = this.deliveryStream.env.region;

    // Following example from https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/SubscriptionFilters.html#FirehoseExample
    // Create a role to be assumed by CWL that can write to this Firehose.
    const role = new iam.Role(scope, 'CloudWatchLogsCanPutRecordsIntoKinesisFirehose', {
      assumedBy: new iam.ServicePrincipal(`logs.${region}.amazonaws.com`),
    });

    this.deliveryStream.grantPutRecords(role);

    return {
      arn: this.deliveryStream.deliveryStreamArn,
      role,
    };
  }
}
