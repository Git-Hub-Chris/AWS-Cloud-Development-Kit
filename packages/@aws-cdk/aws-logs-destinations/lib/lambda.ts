import * as iam from '@aws-cdk/aws-iam';
import * as lambda from '@aws-cdk/aws-lambda';
import * as logs from '@aws-cdk/aws-logs';
import { Construct } from '@aws-cdk/core';

/**
 * Options that may be provided to LambdaDestination
 */
export interface LambdaDestinationOptions{
  /** Whether or not to add Lambda Permissions.
   * @default true
   */
  readonly addPermissions?: boolean | undefined;
}

const defaultLambdaDestinationOptions: LambdaDestinationOptions = {
  addPermissions: true,
};

/**
 * Use a Lambda Function as the destination for a log subscription
 */
export class LambdaDestination implements logs.ILogSubscriptionDestination {
  /**  LambdaDestinationOptions */
  options: LambdaDestinationOptions;
  constructor(private readonly fn: lambda.IFunction, options?: LambdaDestinationOptions) {
    this.options = setLambdaDestinationOptions(options);
  }

  public bind(scope: Construct, logGroup: logs.ILogGroup): logs.LogSubscriptionDestinationConfig {
    const arn = logGroup.logGroupArn;
    if (this.options.addPermissions === true) {
      this.fn.addPermission('CanInvokeLambda', {
        principal: new iam.ServicePrincipal('logs.amazonaws.com'),
        sourceArn: arn,
        // Using SubScription Filter as scope is okay, since every Subscription Filter has only
        // one destination.
        scope,
      });
    }
    return { arn: this.fn.functionArn };
  }
}
function setLambdaDestinationOptions(options: LambdaDestinationOptions | undefined) {
  if (options === undefined) {
    return defaultLambdaDestinationOptions;
  }
  let addPermissions;
  if (options.addPermissions===undefined) {
    addPermissions = true;
  }

  const returnOptions: LambdaDestinationOptions = {
    addPermissions: addPermissions,
  };
  return returnOptions;
}