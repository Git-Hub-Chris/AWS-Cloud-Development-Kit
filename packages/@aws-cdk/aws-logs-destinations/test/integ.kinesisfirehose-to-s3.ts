#!/usr/bin/env node
import * as firehose from '@aws-cdk/aws-kinesisfirehose';
import * as firehoseDestinations from '@aws-cdk/aws-kinesisfirehose-destinations';
import * as logs from '@aws-cdk/aws-logs';
import * as s3 from '@aws-cdk/aws-s3';
import * as cdk from '@aws-cdk/core';
import { Construct } from 'constructs';
import * as dests from '../lib';

/*
 * Stack verification steps:
 * 1. `aws logs create-log-stream --log-group-name <log-group-name> --log-stream-name MyIntegTestLogStream`
 * 2. `aws logs put-log-events --log-group-name <log-group-name> --log-stream-name MyIntegTestLogStream --log-events timestamp=$(date +%s000),message=hello`
 * 3. Wait a few minutes for Kinesis to deliver log to S3
 * 4. `aws s3 ls s3://<bucket-name> --recursive` should list one object
 */

class TestStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // Create Firehose Stream that will deliver to S3 bucket
    const bucket = new s3.Bucket(this, 'Bucket', {
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    });

    const deliveryStream = new firehose.DeliveryStream(this, 'MyFirehose', {
      destinations: [new firehoseDestinations.S3Bucket(bucket)],
    });

    const logGroup = new logs.LogGroup(this, 'LogGroup', {
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    });

    // TODO: For some reason creation of this resource fails: `Subscription (Subscription391C9821) destinationArn for vendor firehose cannot be used with roleArn (Service: AWSLogs; Status Code: 400; Error Code: InvalidParameterException; Request ID: 0e598426-5fcb-4fde-b9d3-11b14c129eb6; Proxy: null)`
    new logs.SubscriptionFilter(this, 'Subscription', {
      logGroup,
      destination: new dests.KinesisFirehoseDestination(deliveryStream),
      filterPattern: logs.FilterPattern.allEvents(),
    });
  }
}

const app = new cdk.App();
new TestStack(app, 'cdk-logs-destinations-firehose-to-s3');
app.synth();
