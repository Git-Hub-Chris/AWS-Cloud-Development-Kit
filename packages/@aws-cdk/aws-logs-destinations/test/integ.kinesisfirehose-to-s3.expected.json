{
  "Resources": {
    "Bucket83908E77": {
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
      "DeletionPolicy": "Delete"
    },
    "FirehoseRoleAA67C190": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "firehose.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": [
                    "s3:AbortMultipartUpload",
                    "s3:GetBucketLocation",
                    "s3:GetObject",
                    "s3:ListBucket",
                    "s3:ListBucketMultipartUploads",
                    "s3:PutObject"
                  ],
                  "Effect": "Allow",
                  "Resource": [
                    {
                      "Fn::GetAtt": [
                        "Bucket83908E77",
                        "Arn"
                      ]
                    },
                    {
                      "Fn::Join": [
                        "",
                        [
                          {
                            "Fn::GetAtt": [
                              "Bucket83908E77",
                              "Arn"
                            ]
                          },
                          "/*"
                        ]
                      ]
                    }
                  ]
                }
              ],
              "Version": "2012-10-17"
            },
            "PolicyName": "allowToDeliverToS3Bucket"
          }
        ]
      }
    },
    "MyFirehose": {
      "Type": "AWS::KinesisFirehose::DeliveryStream",
      "Properties": {
        "ExtendedS3DestinationConfiguration": {
          "BucketARN": {
            "Fn::GetAtt": [
              "Bucket83908E77",
              "Arn"
            ]
          },
          "RoleARN": {
            "Fn::GetAtt": [
              "FirehoseRoleAA67C190",
              "Arn"
            ]
          }
        }
      }
    },
    "MyFirehoseCloudWatchLogsCanPutRecordsIntoKinesisFirehose30DECEBA": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": {
                  "Fn::Join": [
                    "",
                    [
                      "logs.",
                      {
                        "Ref": "AWS::Region"
                      },
                      ".amazonaws.com"
                    ]
                  ]
                }
              }
            }
          ],
          "Version": "2012-10-17"
        }
      }
    },
    "MyFirehoseCloudWatchLogsCanPutRecordsIntoKinesisFirehoseDefaultPolicyF5730531": {
      "Type": "AWS::IAM::Policy",
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "firehose:PutRecord",
                "firehose:PutRecordBatch"
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "MyFirehose",
                  "Arn"
                ]
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "PolicyName": "MyFirehoseCloudWatchLogsCanPutRecordsIntoKinesisFirehoseDefaultPolicyF5730531",
        "Roles": [
          {
            "Ref": "MyFirehoseCloudWatchLogsCanPutRecordsIntoKinesisFirehose30DECEBA"
          }
        ]
      }
    },
    "LogGroupF5B46931": {
      "Type": "AWS::Logs::LogGroup",
      "Properties": {
        "RetentionInDays": 731
      },
      "UpdateReplacePolicy": "Delete",
      "DeletionPolicy": "Delete"
    },
    "Subscription391C9821": {
      "Type": "AWS::Logs::SubscriptionFilter",
      "Properties": {
        "DestinationArn": {
          "Fn::GetAtt": [
            "MyFirehose",
            "Arn"
          ]
        },
        "FilterPattern": "",
        "LogGroupName": {
          "Ref": "LogGroupF5B46931"
        },
        "RoleArn": {
          "Fn::GetAtt": [
            "MyFirehoseCloudWatchLogsCanPutRecordsIntoKinesisFirehose30DECEBA",
            "Arn"
          ]
        }
      }
    }
  }
}
